# Explaining Buckaroo to my friends and family

I have been building the Buckaroo table widget for a year. This is an explainer for my less techincal friends and family members.
In this document I walk through the PyData ecosystem that Bukcaroo exists in, and explain how Buckaroo improves the workflow for users their.

## How programming languages do math
Bear with me, I need to lay some groundwork before we get to exciting charts, graphs, tables and interactive programs.  This whole demo takes 5 minutes.

For most commonly used programming environments like python, javascript, and excel mathematical operations are relatively slow. Everytime a langugae evaluates `c = a + b`.  The computer must first figure out is `a` an integer, decimal point number or string.  Then it must figure out the same for `b`, the it figures out the correct way to add these two types, and assigns the result to `c`.

## Matrix math, why it matters and how NumPy makes it fast.

Matrix math is at the core of linear regression, image recognition, recommendation systems, and AI like chatGPT.  Performing operations on matrices very quickly is at the core of each of these technogies.  In a regular programming language if you wanted to add two 1000x1000 2 dimensional arrays (1_00_000 elements each), for each element the programming language needs to figure out the type, then add them together.

In 2005 Travis Oliphant created NumPy which is built to deal with arrays and matrices.  NumPy determines the type of the whole first matrix, then the type of the whole second matrix, then iterates over the 10_000 elements and adds them together.  This is much much faster... I will show you

In [29]:
ELEMENTS = 1_000_000
py_a = [x for x in range(ELEMENTS)]
py_b = [x for x in range(ELEMENTS*10, 0, -10)]
%timeit -n 10 [py_a[i] + py_b[i] for i in range(ELEMENTS)] 

37.8 ms ± 1.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [30]:
import numpy as np
np_a = np.arange(ELEMENTS)
np_b = np.arange(ELEMENTS * 10, 0, -10)
%timeit -n 10 np_a + np_b

895 µs ± 167 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


# How much faster
37.8 milliseconds vs 895 microseconds, or about 50 times faster. A microsecond is a milionth of a second, a milliseconds is a thousandth